src/encoding/encoding.ts

```diff
@@ -63,5 +72,145 @@ export function encode(obj: Record<string | number | symbol, any>) {
+
+export function msgpackEncodingDataToJSONEncodingData(
+  e: MsgpackEncodingData
+): JSONEncodingData {
+  if (e === null || e === undefined) {
+    return e;
+  }
+  if (e instanceof Uint8Array) {
+    return bytesToBase64(e);
+  }
+  if (Array.isArray(e)) {
+    return e.map(msgpackEncodingDataToJSONEncodingData);
+  }
+  if (e instanceof Map) {
+    const obj: { [key: string]: JSONEncodingData } = {};
+    for (const [k, v] of e) {
+      if (typeof k !== 'string') {
+        throw new Error(`JSON map key must be a string: ${k}`);
+      }
+      obj[k] = msgpackEncodingDataToJSONEncodingData(v);
+    }
+    return obj;
+  }
+  return e;
+}
+
+export function jsonEncodingDataToMsgpackEncodingData(
+  e: JSONEncodingData
+): MsgpackEncodingData {
+  if (e === null || e === undefined) {
+    return e;
+  }
+  if (
+    typeof e === 'string' || // Note, this will not convert base64 to Uint8Array
+    typeof e === 'number' ||
+    typeof e === 'bigint' ||
+    typeof e === 'boolean'
+  ) {
+    return e;
+  }
+  if (Array.isArray(e)) {
+    return e.map(jsonEncodingDataToMsgpackEncodingData);
+  }
+  if (typeof e === 'object') {
+    const obj = new Map<string, MsgpackEncodingData>();
+    for (const [key, value] of Object.entries(e)) {
+      obj.set(key, jsonEncodingDataToMsgpackEncodingData(value));
+    }
+    return obj;
+  }
+  throw new Error(`Invalid JSON encoding data: ${e}`);
+}
+
+/* eslint-disable class-methods-use-this */
+/* eslint-disable no-useless-constructor,no-empty-function */
+
+export abstract class Schema {
+  public abstract defaultValue(): unknown;
+
+  public abstract isDefaultValue(data: unknown): boolean;
+
+  public abstract prepareMsgpack(data: unknown): MsgpackEncodingData;
+
+  public abstract fromPreparedMsgpack(encoded: MsgpackEncodingData): unknown;
+
+  public abstract prepareJSON(data: unknown): JSONEncodingData;
+
+  public abstract fromPreparedJSON(encoded: JSONEncodingData): unknown;
+}
+
+export interface Encodable {
+  toEncodingData(): unknown;
+  getEncodingSchema(): Schema;
+}
+
+export interface EncodableClass<T extends Encodable> {
+  fromEncodingData(data: unknown): T;
+  encodingSchema: Schema;
+}
+
+export function decodeMsgpack<T extends Encodable>(
+  encoded: ArrayLike<number>,
+  c: EncodableClass<T>
+): T {
+  return c.fromEncodingData(
+    c.encodingSchema.fromPreparedMsgpack(
+      decodeAsMap(encoded) as MsgpackEncodingData
+    )
+  );
+}
+
+export function encodeMsgpack(e: Encodable): Uint8Array {
+  return rawEncode(e.getEncodingSchema().prepareMsgpack(e.toEncodingData()));
+}
+
+export function decodeJSON<T extends Encodable>(
+  encoded: string,
+  c: EncodableClass<T>
+): T {
+  const decoded: JSONEncodingData = parseJSON(encoded, {
+    intDecoding: IntDecoding.MIXED,
+  });
+  return c.fromEncodingData(
+    c.encodingSchema.fromPreparedJSON(decoded) as JSONEncodingData
+  );
+}
+
+export function encodeJSON(e: Encodable, space?: string | number): string {
```

Comment:
Where would the encode/decode JSON testing be found? (Do find some direct calls to the prepareJSON function off the Schema base class)